### ISSUE TITLE
semantics change, faster bulk eviction

### ISSUE THREAD

**Issue by isaacs** [isaacs/node-lru-cache#186]
tl;dr - use a generational map algorithm instead of tracking precise use recency of each item in cache.

- Generations should be `max` size.  We don't save anything by making them smaller, since the max possible items stored is still the same, unless we _also_ limit generation size on `get()`, which is too much object creation.
- instead of caching _exactly_ the `max` least recently used items, we cache _at least_ `max` least recently used items
- total cache size _may_ grow to up to `2*max-1` items, so tune appropriately.

---

exploration below...

Right now, the algorithm is ok, but it's not as efficient as it could be, and doesn't leverage the things that JS is good at.  If we loosen the restriction that `cache.length` will strictly never exceed `cache.max`, we can do some interesting things.

We don't have to keep shuffling around a list or a counter to keep track of each mru value, fewer indexes, fewer `delete` calls on big JavaScript objects, etc.  Just a parent reference that gets set to `null` and garbage collected.

Create an internal `Cache` class that has:

- a map of key => {value,expiration}
- a parent `Cache`
- length of items in cache, which includes parent items `get length () { return this.map.size + (this.parent ? this.parent.length : 0) }`
- a prune method that is `this.parent = null`

`Cache.has` looks like:

```js
if (this.map.has(key)) {
  return true
} else if (this.parent) {
  return this.parent.has(key)
} else {
  return false
}
```

`Cache.get` looks like:

```js
if (this.map.has(key)) {
  return this.map.get(key)
} else {
  for (let parent = this.parent; parent; parent = parent.parent) {
    if (parent.map.has(key)) {
      const value = parent.map.get(key)
      this.map.set(key, value)
      return value
    }
  }
}
return undefined
```

So we're always pulling items up to the youngest `Cache` object as they're used, only adding a few `Map.get()` calls, which are very fast.

Cache.set just calls `this.map.set(key, value)`

Actually `Cache.prune` would be more like:

```js
prune (max) {
  if (this.map.size >= max) {
    // have enough locally, drop parent
    this.parent = null
  }
  if (this.parent) {
    this.parent.prune(max - this.map.size)
  }
}
```

Whenever `cache.map.length` is at 50% of the lru max size, we create a new generation.  So, if you have a lru max of 100, then when you add the 51st item, we create a new child cache.  The parent can thus never be greater than length of 50.  As you use items, those are pulled into the child cache.  When _it_ gets to 50 items in that most recent generation, we create a new generation.

When the `cache.length` is >= max * 1.5, we call `prune()`, which will lop off earlier generations.

So, say we have a max of 10, and keep adding items:

- add 5 items, one cache generation `{1,2,...,5}` len=5
- add item 6-10, in new cache generation, so we have `{6,7,...,10} -> {1,2,...,5}` len=10
- add item 11-15, we have `{11,12,...,15}->{6,7,...,10}->{1,2,...,5}` len=15
- at this point, `len >= max*1.5`, so we prune, and find we can drop the oldest generation `{11,12,...,15}->{6,7,...,10}`, len=10

The `50%` and `1.5` can be an option.  Basically, with some `x` <= 1, we create new generations when we hit `Math.ceil(max*x)` and we prune at `Math.ceil(max*(1+x))`.  Higher `x` factor will mean more speed, but also more memory usage.  At `x=1`, you basically will never prune until the current generation is at the max length, and may be using up to `max * 2` storage.  As `x` approaches 0, you'll be creating new `Map` objects for every new entry, and pruning every time it touches the ceiling.  (Most slow and memory efficient.)

Another scenario where we `get` items periodically, x=0.5, max=10

- add 5 items, one cache generation `{1,2,...,5}` len=5
- add 3 items, new gen, `{6,7,8}->{1,2,...,5}` len=8
- get(5) `{5,6,7,8}->{1,2,3,4}` len=8
- get(4) `{4,5,6,7,8}->{1,2,3}` len=8
- add 2 more items, but these go in a new generation, because size>=5 `{9,10}->{4,5,6,7,8}->{1,2,3}` len=10
- add 5 more items, new gen `{14,15}->{9,10,11,12,13}->{4,5,6,7,8}->{1,2,3}` len=15
- now length hits max*1.5, so we `prune(10)` on the youngest generation
    - 4th gen size is 2, < 10, so don't drop.  prune parent to 8 `(max-this.map.length)`
    - 3rd gen size is 5, < 8, so don't drop.  prune parent to 3
    - 2nd gen size is 5, >=3, so drop parent.
    - result: `{14,15}->{9,10,11,12,13}->{4,5,6,7,8}` len=12
- get(4): `{4,14,15}->{9,10,...,13}->{5,6,7,8}` len=12
- get(5): `{4,5,14,15}->{9,10,11,12,13}->{6,7,8}` len=12
- get(6): `{4,5,6,14,15}->{9,10,11,12,13}->{7,8}` len=12
    - at this point, since we went all the way to the oldest ancestor, we can note that pruning that ancestor still keeps us above target max, and cut it off
    - result: `{4,5,6,14,15}->{9,10,11,12,13}` len=10
- get(7): not found in cache, `undefined`
- get(9), `{4,5,6,9,14,15}->{10,11,12,13}`
- get(10), this spawns a new gen, because `map.size>=max*1.5`: `{10}->{4,5,6,9,14,15}->{11,12,13}`

That last point isn't so great?  It means that if you have roughly `max` items, and you keep requesting them, you'll end up creating and throwing away `Map` objects repeatedly.  Maybe it'd be better if new gens are created on `get` _only_ if it would bring the youngest generation up to `max` size.  Picking up from there...

- get(10), does not spawn a new gen, because `map.size < max`: `{4,5,6,9,10,14,15}->{11,12,13}` len=10
- put(16), _does_ spawn a new gen, because `map.size >=max*1.5`: `{16}->{4,5,6,9,10,14,15}->{11,12,13}` len=11
- get(11): `{16,11}->{4,5,6,9,10,14,15}->{12,13}`, len=11
- get(12): `{16,11,12}->{4,5,6,9,10,14,15}->{13}`, len=11
     - walked all the way to last gen, and removing it drops us to max, so drop it: `{16,11,12}->{4,5,6,9,10,14,15}`, len=10
- get(13): `undefined`
- get(4):  `{4,12,16,17}->{5,6,9,10,14,15}` len=10
- get(5):  `{4,5,12,16,17}->{6,9,10,14,15}` len=10
- get(6):  `{4,5,6,12,16,17}->{9,10,14,15}` len=10
- get(9):  `{4,5,6,9,12,16,17}->{10,14,15}` len=10
- get(10):  `{4,5,6,9,10,12,16,17}->{14,15}` len=10
- get(14):  `{4,5,6,9,10,14,12,16,17}->{15}` len=10
- get(15):  `{4,5,6,9,10,14,15,12,16,17}->{}` len=10
    - at this point, we can drop the last gen and stay above max (10-0=10), so drop it:`{4,5,6,9,10,14,15,12,16,17}`
    - now we have a single generation with `max` items in it.
- put(18): `{18}->{4,5,6,9,10,14,15,12,16,17}` len=11
- put(19): `{18,19}->{4,5,6,9,10,14,15,12,16,17}` len=12
- put(20): `{18,19,20}->{4,5,6,9,10,14,15,12,16,17}` len=13
- put(21): `{18,19,20,21}->{4,5,6,9,10,14,15,12,16,17}` len=14
- put(22): `{18,19,20,21,22}->{4,5,6,9,10,14,15,12,16,17}` len=15
- put(23): `{23}->{18,19,20,21,22}->{4,5,6,9,10,14,15,12,16,17}` len=16
- put(24): `{23,24}->{18,19,20,21,22}->{4,5,6,9,10,14,15,12,16,17}` len=17
- put(25): `{23,24,25}->{18,19,20,21,22}->{4,5,6,9,10,14,15,12,16,17}` len=18
- put(26): `{23,24,25,26}->{18,19,20,21,22}->{4,5,6,9,10,14,15,12,16,17}` len=19
- put(27): `{23,24,25,26,27}->{18,19,20,21,22}->{4,5,6,9,10,14,15,12,16,17}` len=20
    - at this point, we can drop the oldest generation, leaving `{23,24,25,26,27}->{18,19,20,21,22}` len=10

Let's back up to before `put(27)` and see what happens if we pull entries up to the youngest generation from the middle generation:

- put(26): `{23,24,25,26}->{18,19,20,21,22}->{4,5,6,9,10,14,15,12,16,17}` len=19
- get(18):  `{18,23,24,25,26}->{19,20,21,22}->{4,5,6,9,10,14,15,12,16,17}` len=19
- get(19):  `{18,19,23,24,25,26}->{20,21,22}->{4,5,6,9,10,14,15,12,16,17}` len=19
- get(20):  `{18,19,20,23,24,25,26}->{21,22}->{4,5,6,9,10,14,15,12,16,17}` len=19
- get(21):  `{18,19,20,21,23,24,25,26}->{22}->{4,5,6,9,10,14,15,12,16,17}` len=19
- get(22):  `{18,19,20,21,22,23,24,25,26}->{}->{4,5,6,9,10,14,15,12,16,17}` len=19
    - Any time `Cache.parent.map.size === 0`, we set `this.parent = this.parent.parent` to splice it out, so:
    - `{18,19,20,21,22,23,24,25,26}->{4,5,6,9,10,14,15,12,16,17}` len=19
- The next `put` or `get` will cause the parent gen to be dropped
    - put: length is 20, oldest generation is 10, drop the 10 and come down to max
    - get: youngest generation is 10 items, drop its parent and come down to max

Back up to before `put(27)` and see what happens if we pull entries from the oldest generation into the youngest.

- put(26): `{23,24,25,26}->{18,19,20,21,22}->{4,5,6,9,10,14,15,12,16,17}` len=19
- get(4):  `{4,23,24,25,26}->{18,19,20,21,22}->{5,6,9,10,14,15,12,16,17}` len=19
    - at this point, we can prune the oldest generation out, because it's 9 and removing it drops us to 10
    - `{4,23,24,25,26}->{18,19,20,21,22}` len=10

How about if the generational coefficient isn't 0.5, but instead something like 0.2?  Max still 10, so generations will grow to between 2 and 10 items per map, and let's try to get the size usage as high as possible.

- put(1): `{1}`
- put(2): `{1,2}`
- put(3): `{3}->{1,2}`
- put(4): `{3,4}->{1,2}`
- put(5): `{5}->{3,4}->{1,2}`
- put(6,7,8,9,10): `{9,10}->{7,8}->{5,6}->{3,4}->{1,2}` len=10
- get(1,2,3,4,5,6,7,8): `{1,2,3,4,5,6,7,8,9,10}->{}`, drop empty map, `{1,2,3,4,5,6,7,8,9,10}`
- put(11): `{11}->{1,2,3,4,5,6,7,8,9,10}` len=11
- put(12): `{11,12}->{1,2,3,4,5,6,7,8,9,10}` len=12
- put(13): `{13}->{11,12}->{1,2,3,4,5,6,7,8,9,10}` len=13
- put(14): `{13,14}->{11,12}->{1,2,3,4,5,6,7,8,9,10}` len=14
- put(15,16,17,18,19): `{19}->{17,18}->{15,16}->{13,14}->{11,12}->{1,2,3,4,5,6,7,8,9,10}` len=19
    - any get of items in the oldest generation will make it drop, because 19-9=10
    - any new put will also make it drop, because 20-10=10
    - let's get all the items that _aren't_ in the oldest generation
- get(11,12,13,14,15,16,17,18): `{11,12,13,14,15,16,17,18,19}->{1,2,3,4,5,6,7,8,9,10}` len=19
- at this point, any get or put will cause the older gen to be dropped.

Same thing, but generational coefficient is 0.1 (so 1 item per map before a new gen is spawned)

- put(1): `{1}`
- put(2): `{2}->{1}`
- put(3): `{3}->{2}->{1}`
- put(4,5,6,7,8,9,10): `{10}->{9}->{8}->{7}->{6}->{5}->{4}->{3}->{2}->{1}` len=10
    - at this point, any get() less than 10 will drop a generation, and any put() will cause the 1 to be dropped
- get(1,2,3,4,5,6,7,8,9): `{1,2,3,4,5,6,7,8,9,10}->{}`, drop empty maps along the way, same as before, `{1,2,3,4,5,6,7,8,9,10}`
- put(11): `{11}->{1,2,3,4,5,6,7,8,9,10}` len=11
- put(12): `{12}->{11}->{1,2,3,4,5,6,7,8,9,10}` len=12
- put(13,14,15,16,17,18,19): `{19}->{18}->{17}->{16}->{15}->{14}->{13}->{12}->{11}->{1,2,3,4,5,6,7,8,9,10}` len=19
    - any get of items in the oldest generation will make it drop, because 19-9=10
    - any new put will also make it drop, because 20-10=10
    - let's get all the items that _aren't_ in the oldest generation
- get(11,12,13,14,15,16,17,18): `{11,12,13,14,15,16,17,18,19}->{1,2,3,4,5,6,7,8,9,10}` len=19
- at this point, any get or put will cause the older gen to be dropped.

So the same as with the generation size at 0.5.  Max total size is `2*max-1`

What if the generation size is 1, so we can store 10 per generation?
- put(1): `{1}`
- put(2): `{2}->{1}`
- put(3): `{3}->{2}->{1}`
- put(1,2,3,4,5,6,7,8,9,10): `{1,2,3,4,5,6,7,8,9,10}` len=10
- get(1,2,3,4,5,6,7,8,9): `{1,2,3,4,5,6,7,8,9}->{10}`, but no empty maps along the way
    - any `put` here will make the oldest gen drop
- get(10), that pulls it back into the top gen: `{1,2,3,4,5,6,7,8,9,10}` len=10
- put(11): `{11}->{1,2,3,4,5,6,7,8,9,10}` len=11
- put(12,13,14,15,16,17,18,19): `{11,12,13,14,15,16,17,18,19}->{1,2,3,4,5,6,7,8,9,10}` len=19
    - any get of items in the older generation will make it drop, because 19-9=10
        - eg: get(10): `{10}->{11,12,13,14,15,16,17,18,19}->{1,2,3,4,5,6,7,8,9}` len=19, drop the last gen for `{10}->{11,12,13,14,15,16,17,18,19}` len=10
    - any new put will also make it drop, because 20-10=10

So I don't know how much value there really is in limiting the size of the generations, since they can grow on `get()` up to `max` anyway, or else we're shuffling around an awful lot on rapid-fire `get()` calls, which is the most common use case for any cache.  That would be useful if we didn't mind having items be stored in multiple generations.  Because then we could just say "we will track `max` indexes" and always _copy_ them into the younger maps, and drop to keep the size below some fixed amount.  Then you want smaller generations to have better granularity in how close your index count is to the actual lru "max items cached" that the user will think about, but even then, I feel like it's too big a divergence in semantics, whereas this (you might get an item less recently used than the top `max` most recently used items, but no more recently used than the top `2*max-1` items) is a much smaller step to take.

Another advantage of making the generational limit the max is that we will be guaranteed to only ever have 2 generations.  Before it would ever cause a 3rd generation to be spawned, we'd just drop the oldest one to drop down to `max`.  The object lookups and `map.size` calculations are going to be the slowest part of all this, so keeping the tree short is key.  Effectively, we just have "current" and "old" generations, and drop "old" when "current" reaches max size.

Bottom line:
- Generations should be `max` size.  We don't save anything by making them smaller, since the max possible items stored is still the same, unless we _also_ limit generation size on `get()`, which is too much object creation.
- instead of caching _exactly_ the `max` least recently used items, we cache _at least_ `max` least recently used items
- total cache size _may_ grow to up to `2*max-1` items.

**Comment by isaacs** [comment]
Even simpler approach, actually, since we can only ever have at most 2 gens:

```
class Entry {
  constructor (value) {
    this.value = value
  }
}

class Cache {
  constructor (max) {
    this.max = max
    this.old = new Map()
    this.current = new Map()
  }
  get size () {
    return this.current.size + this.old.size
  }
  set (key, value) {
    this.current.set(key, new Entry(value))
    this.prune()
  }
  get (key) {
    const fromCurrent = this.current.get(key)
    if (fromCurrent) {
      return fromCurrent.value
    } else {
      const fromOld = this.old.get(key)
      if (fromOld) {
        this.current.set(key, fromOld)
        this.prune()
        return fromOld.value
      }
    }
  }
  prune () {
    if (this.current.size >= this.max) {
      this.old = this.current
      this.current = new Map()
    }
  }
}
```

The real sticky wicket in all of this is handling `maxAge` and entry ttl...
